txnsync/profiler.go

```diff
@@ -0,0 +1,205 @@
+// Copyright (C) 2019-2021 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package txnsync
+
+import (
+	"time"
+
+	"github.com/algorand/go-algorand/logging"
+	"github.com/algorand/go-algorand/logging/telemetryspec"
+	"github.com/algorand/go-algorand/util/timers"
+)
+
+//msgp:ignore profElements
+type profElements int
+
+const (
+	profElementIdle = iota
+	profElementTxChange
+	profElementNewRounnd
+	profElementPeerState
+	profElementIncomingMsg
+	profElementOutgoingMsg
+	profElementNextOffset
+
+	// detached elements
+	profElementGetTxnsGroups
+	profElementAssembleMessage
+	profElementSendMessage
+	profElementMakeBloomFilter
+	profElementTxnsSelection
+
+	profElementLast
+	profFirstDetachedElement = profElementGetTxnsGroups
+)
+
+// The profiler struct provides profiling information regarding the main loop performance
+// characteristics. Using it provides statistics information about the recent duty cycle utilization,
+// that could be used when trying to throlle the accuracy and performance of the transaction sync.
+type profiler struct {
+	// clock used as the source clock for measurements.
+	clock timers.WallClock
+	// elements contains the elements we want to measure.
+	elements []*element
```

Comment:
I'm guessing the main purpose of `elements` is to reduce the amount of boilerplate code. How about having multiple public variables here for each element and having `profile` contain pointers to one of the elements? Then `prune` can be as concise as it is now and the user of `profiler` can also access elements without any extra code. I think it will be just as concise.

My main problem with `elements` is that it makes readability worse. It took me some time to understand the code.

Reply:
I think that having all these defined on the object itself is quite an ugly implementation. I was trying to create a simpler struct interface, and enumerate the different elements using a simple constants instead.
Having said that, I do agree that we could do better with the documentation here so that it would be more clear why we chose to do that in this way. I'll try to improve the docs, but please don't be shy, and if you'd like to suggest a specific description, please let me know.

Reply:
I guess it's not important right now. When this code needs to be merged to master, it can be revisited.